web: add web-app server for development and production builds

valerybugakov opened the pull request:
# Independent web app server

This PR introduces an independent server for the web application which can use any deployed API instance as a backend.

## Context

While working on the CSS modules configuration, I needed a way to iterate quickly on the `webpack.config.js` and test changes in the production environment: run `yarn build-web` and check how CSS modules are loaded in the application. There was no easy/quick way to do it. @umpox found out that it's possible [to change a couple of variables](https://github.com/sourcegraph/sourcegraph/commit/c5c21faf4051a54ae9e2d3b2440c4bd43f9caf8e) in the `start.sh` script to serve the production build. But this solution was quite slow because of the backend services involved. 

To solve this, I've configured a minimum viable server to serve web app production artifacts without our backend services. I've added [HtmlWebpackPlugin](https://v4.webpack.js.org/plugins/html-webpack-plugin/) to our `webpack.config.js` to produce `index.html` and configured [Webpack dev server proxy](https://webpack.js.org/configuration/dev-server/#devserverproxy) to point the server to the deployed API instance. It allowed testing changes in production build without backend services quickly. This PR is the enhanced version of this setup.

## How it works

1. [HtmlWebpackPlugin](https://v4.webpack.js.org/plugins/html-webpack-plugin/) is added to the Webpack config. It can be enabled by the env variable `WEBPACK_SERVE_INDEX=true`. This plugin outputs `index.html` to the `ui/assets`.
2. Express server is used for a production build. Webpack-dev-server in the development environment.
2. Server proxies API requests to `SOURCEGRAPH_API_URL` provided in the `.env` file using the [http-proxy-middleware](https://github.com/chimurai/http-proxy-middleware). 
3. To avoid the `CSRF token is invalid` API error CSRF token is retrieved from the `SOURCEGRAPH_API_URL` before the server starts. Then this value is used for every subsequent request to the API.
4. To enable authentication via the `sgs` cookie, all cookies from the API requests are propagated to the local client.

## Known issues

The issues listed below can be addressed in separate PRs. They are non-blocking for the functionality this PR introduces and doesn't affect existing workflows. We can iterate on each individual issue once we have the foundation in place. 

1. HTTPS support is not in place yet, because of (https://github.com/webpack/webpack-dev-server/issues/2313). It needs more investigation.
2. A static mock in combination with a local `site-config.json` is used instead of the JS context returned from the `SOURCEGRAPH_API_URL`. 
2. Authentication providers don't work. This is probably related to the social login links, which are served along with JS context. 
3. A couple of variables are duplicated from `gulpfile.js`. They should be shared with it after migration to Typescript.
5. `index.html` template is different from what we have in `cmd/frontend/internal/app/ui/app.html`. We should reuse it to be consistent with the default production setup.

## Changes

1.   Two commands are added to the `client/web/package.json`: `yarn serve:dev` and `yarn serve:prod`.
2. `dev` directory is added to the `client/web` package. It contains the server configuration used by the commands above.
3. `.env.example` is added to the `client/web` to showcase the example env configuration.
4. `require('ts-node').register()` is added to the `gulpfile.js` to allow TS usage for `html-webpack-plugin` configuration.

## Updates

1. Added four commands to integrate web server with `sg` CLI tool:
  - `sg run web-standalone`
  - `sg run enterprise-web-standalone`
  - `sg run web-standalone-prod`
  - `sg run enterprise-web-standalone-prod`
2. Removed `.env.example` and `dotenv` package because `env` variables are now defined in `sg.config.yaml`.
3. Added global alert about API proxy usage.

<img width="1193" alt="Screenshot 2021-05-05 at 14 15 12" src="https://user-images.githubusercontent.com/3846380/117132996-3e9cc480-add6-11eb-868d-dee5101526ea.png">

eseliger reviewed (approved):
Nice work 🎊 

I would be worried a bit about people messing with sourcegraph.com, but maybe that's not a reasonable fear :) I think we should not make it serve on sourcegraph.test:3443, though because it's too easy to think it's a safe dev environment.

- A static mock is used instead of the JS context returned from the SOURCEGRAPH_API_URL.

I think that should be implemented soon. The flags set change a lot about the app behavior and it is also dynamic in the backend so if I change a config, the jscontext _would_ change if it was a real deployed app.

- index.html template is different from what we have in cmd/frontend/internal/app/ui/app.html. We should reuse it to be consistent with the default production setup.

Definitely agree. Also it's so easy to forget about this file when someone updates the `app.html` file.

felixfbecker commented on client/web/dev/utils/get-site-config.ts:
```diff
@@ -0,0 +1,26 @@
+export const getSiteConfig = (): Partial<SourcegraphContext> => {
+    try {
+        // eslint-disable-next-line no-sync
+        const siteConfig = JSON.parse(stripJsonComments(fs.readFileSync(SITE_CONFIG_PATH, 'utf-8')))
```
+JSONC can also contain trailing commas

valerybugakov commented on client/web/dev/utils/get-api-proxy-settings.ts:
```diff
@@ -0,0 +1,42 @@
+    apiURL: string
+}
+
+export const getAPIProxySettings = ({ csrfContextValue, apiURL }: GetAPIProxySettingsOptions): Options => ({
```
yes, it should work too

valerybugakov commented on client/web/README.md:
```diff
@@ -0,0 +1,31 @@
+
+### Prerequisites
+
+1. Duplicate `client/web/.env.example` as `client/web/.env`.
```
right, changed this title to `configuration`

felixfbecker commented on client/web/dev/server/development.server.ts:
```diff
@@ -0,0 +1,65 @@
+
+    const options: WebpackDevServer.Configuration = {
+        hot: IS_HOT_RELOAD_ENABLED,
+        // TODO: resolve https://github.com/webpack/webpack-dev-server/issues/2313 and enable HTTPS.
```
The linked issue is closed, could you give more details?

valerybugakov replied:
Webpack dev server still shows the same error as described in this issue with `https: true` in configuration. There're multiple issues from other repos which mention that they've disabled HTTPS support because of it. I didn't dig too deep with this one, that's all context for the moment. Just need to allocate more time for this one.

valerybugakov commented on client/web/dev/server/development.server.ts:
```diff
@@ -0,0 +1,65 @@
+import 'dotenv/config'
```
This is the right call. I thought about doing it in this PR. Still, apart from this change, we need to extract shared Webpack configuration into a standalone package to avoid adding loaders to three different configs: Web, Browser, and Storybook. I believe this step should be done before extracting `client/dev-server` because it might result in duplicated work and circular dependencies between `web` and `dev-server` packages. First, move configs to a shared package. Then extract dev-server and use shared configs.

valerybugakov commented on client/web/dev/utils/get-site-config.ts:
```diff
@@ -0,0 +1,26 @@
+        // eslint-disable-next-line no-sync
+        const siteConfig = JSON.parse(stripJsonComments(fs.readFileSync(SITE_CONFIG_PATH, 'utf-8')))
+
+        return lodash.mapKeys(siteConfig, (_value, key) => lodash.camelCase(key))
```
At least web app consumes it in this format :)

felixfbecker commented:
I'm really excited to enable web dev without needing any backend services.

felixfbecker commented:
Something we may want to consider is injecting some kind of banner somewhere "THIS USES PRODUCTION DATA FROM sourcegraph.com" because it may be easy to forget if you have multiple or old tabs open given this runs on localhost. I feel like it may otherwise only be a matter of time until someone accidentally misconfigures an external service on dotcom or something

valerybugakov commented on client/web/gulpfile.js:
```diff
@@ -1,4 +1,11 @@
 // @ts-check
+const path = require('path')
+
+require('ts-node').register({
+  transpileOnly: true,
+  // Use config with "module": "commonjs" because not all modules involved in tasks are esnext modules.
+  project: path.resolve(__dirname, './dev/tsconfig.json'),
+})
```
It's needed because the `get-html-webpack-plugin` is written in TS and `webpack.config.js` uses `get-html-webpack-plugin`. It can be rewritten in JS to solve this issue with a couple of modules, like `create-js-context`. 

I'm curious about the startup regression. I've looked into it briefly, and definitely, `ts-node` adds some overhead.
I've executed the `generate` gulp command in a loop 100 times with and without `ts-node-transpile-only` and the difference is around 0.2-0.7s per execution. It's not a comprehensive study of the startup time, but I wanted to get some feel for it :) The overhead seems negligible to the command execution time. At the same time, it's not possible to use files written in Typescript if you need them in some build task.

I see the `browser` package uses `ts-node` in an interesting way. The `gulpfile.js` itself is written in JS, so no overhead is added by `ts-node` at this point. But it calls the `yarn build` command, which uses `ts-node/register` to run `scripts/build.ts`. It looks like a direct usage `ts-node-transpile-only myScript.ts` should be good enough for development tasks. Certainly, I might be missing some use-cases that I didn't encounter in the codebase yet. Please let me know if that's the case.

valerybugakov replied:
I will resolve this issue for now. Benefits outweigh the small-ish performance gain based on my tests. And benefits will give more fruits in the future when we have a package for shared config chunks. I am always happy to revisit this question in the future if the regression is noticeable.

valerybugakov commented on client/web/dev/server/development.server.ts:
```diff
@@ -0,0 +1,65 @@
+    WEB_SERVER_URL,
+} from '../utils'
+
+// TODO: migrate webpack.config.js to TS to use `import` in this file.
```
Thanks for raising this, discussing it below in a thread with Felix: https://github.com/sourcegraph/sourcegraph/pull/20126#discussion_r622304304. I will resolve this for now.

umpox reviewed (commented):
After following the READM

vovakulikov commented on client/web/README.md:
````diff
@@ -0,0 +1,31 @@
+### Development server
+
+```sh
+ENTERPRISE=1 yarn serve:dev
````
Hey, I've gotten the same thing and It turned out I forgot to create .end file from .env.example.

vovakulikov replied:
Probably something has been changed and I looked on some first version of now I also have this problem 

@valerybugakov maybe we should include something to load variables to the process from .env file 


```
// development.server.ts
import dotenv from 'dotenv'

dotenv.config()
```

valerybugakov replied:
@vovakulikov I've just updated the PR. See the comment: https://github.com/sourcegraph/sourcegraph/pull/20126#issuecomment-832607838.
`dotenv` package was removed in favor of the `sg` CLI tool that we use to start our services to have one source of truth for env variables. 

The previous version had `import 'dotenv/config'` at the top of each `.server.ts` file.

vovakulikov commented:
Nice work. I don't thing this is a blocker for this, but just thing which we should notice 
for some reason we have strange visual behaviour in safari with this approach. Probably some styles for Monaco editor are missing.

![image](https://user-images.githubusercontent.com/18492575/116856277-89b2ae00-ac03-11eb-9112-2e8fac9a2dd3.png)

valerybugakov commented on client/web/README.md:
````diff
@@ -0,0 +1,31 @@
+
+```sh
+ENTERPRISE=1 NODE_ENV=production DISABLE_TYPECHECKING=true yarn run build
+yarn serve:prod
````
Updated `README.md` to use the `sg` CLI tool.

valerybugakov commented on client/web/README.md:
````diff
@@ -0,0 +1,31 @@
+### Production server
+
+```sh
+ENTERPRISE=1 NODE_ENV=production DISABLE_TYPECHECKING=true yarn run build
````
Extracted everything into `sg` commands. Works like a charm.  👍 
Thanks for a suggestion, @eseliger, and @mrnugget, for building this!

mrnugget replied:
Nice!

mrnugget commented on sg.config.yaml:
```diff
@@ -186,10 +185,53 @@ commands:
+    env:
+      ENTERPRISE: 1
+      NODE_ENV: production
+      NODE_OPTIONS: "--max_old_space_size=4096"
```
FYI: You can use the top-level `env` to define env vars that should be set for all commands/tests. So instead of repeating the `NODE_OPTIONS` or the `HTTPS_DOMAIN`, you can set it once :)

valerybugakov replied:
@mrnugget updated, thanks for the suggestion!

valerybugakov commented:
## Updates

1. Added four commands to integrate web server with `sg` CLI tool:
  - `sg run web-standalone`
  - `sg run enterprise-web-standalone`
  - `sg run web-standalone-prod`
  - `sg run enterprise-web-standalone-prod`
2. Removed `.env.example` and `dotenv` package because `env` variables are now defined in `sg.config.yaml`.
3. Added global alert about API proxy usage.

<img width="1193" alt="Screenshot 2021-05-05 at 14 15 12" src="https://user-images.githubusercontent.com/3846380/117132996-3e9cc480-add6-11eb-868d-dee5101526ea.png">

andreeleuterio reviewed (approved):
LGTM! What matters more for security here (please add anything I didn't think of!):
- Makes devs lives easier ✨
- Uses k8s.sgdev as default
- Alerts users of which instance they're connected to
- No HTTPS _yet_  because of upstream bug
